[Gabrielle Gianna Tan-Winings] iP

gabriellegtw opened the pull request:
# Update Maxine's To-Do List Functionality

[This](https://github.com/gabriellegtw/ip/tree/master) is the link to my repo! 

We’ve made several updates to Maxine, our To-Do list chatbot. Below are the details of the changes:

**Changes Made**
- Improved Task Management: Added features for better task categorization and prioritization.
- Enhanced User Interaction: Updated prompts and responses for a more intuitive user experience.
- Bug Fixes: Fixed several issues related to task reminders and notifications.

_To-Do_

1. More comprehensive user testing: Test that all bugs have been squashed
2. Review User Prompts: Ensure all user prompts are clear and effective.
3. Complete GUI

This is my `main` method:
```java
public static void main(String[] args) {
        new Maxine().run();
    }
```

Tasks 😉

-  [x] Review code changes
-  [x] Update user documentation
-  [x] Perform end-to-end testing

> "The best way to predict the future is to create it." - Peter Drucker

 ~~Notes~~
Maxine is here to help!

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

gabriellegtw committed:
Add greeting, prompt and exit
* Added greeting and exit
* Added prompt to ask user
* Renamed Duke to Maxine

gabriellegtw committed:
Add javadocs
* Added javadocs
* Created method to prompt user

gabriellegtw committed:
Add list function
When user inputs item into the input, the element is added to a set.

When user types "list" (regardless of upper or lowercase), the list will be displayed

gabriellegtw committed:
Add check/uncheck function
* User is able to mark a task as complete using "mark"
* User is able to unmark a task as incomplete using "unmark"

gabriellegtw committed:
Change HashMap to ArrayList
Tasks are stored in ArrayList instead of a hashmap

gabriellegtw committed:
Add logic to handle different task types
* Handle Todo, deadlines, event

gabriellegtw committed:
Add automated testing
* Modified EXPECTED.TXT and ACTUAL.TXT
Note: Ensure that setting is such that whitespaces are not removed on save

gabriellegtw committed:
Add exception handling for different cases

gabriellegtw committed:
Add delete function to delete tasks

gabriellegtw committed:
Improved deletion
* print string to notify which task is deleted

gabriellegtw committed:
Broke up code into methods to enhance readability

gabriellegtw committed:
Change boolean variable names

gabriellegtw committed:
Add maxine.txt and file refresh
List does not store its elements remotely

List needs to store data on another file so that it is able to retrieve the information later.

Let's add a text file and a refresh method so that information can be retrieved

gabriellegtw committed:
Add TaskParser class

When program is launched, list does not read maxine.txt and starts on a new list

Let's add TaskParser class to parse maxine.txt and add to Array list

gabriellegtw committed:
Fix formatting and add tag

gabriellegtw committed:
Handle case where file is not found

File Reading class does not handle FileNotFound Exception

Let's handle the exception within FileReading class internally

gabriellegtw committed:
Delete unused method

gabriellegtw committed:
Merge pull request #1 from gabriellegtw/branch-Level-7

Add automatic list storage in maxine.txt

gabriellegtw committed:
Fix date recognition

The bot takes in the deadline as a string. Taking the string as is may not be very meaningful to the reader

Let's have the bot recognise dates and display it in a different format (eg. typing 18-02-2024 in the input becomes 18th February 2024 in maxine.txt)

gabriellegtw committed:
Merge pull request #2 from gabriellegtw/branch-Level-8

Add date recognition

gabriellegtw committed:
Add Ui class

Print statements within Maxine bot are not encapsulated

Let's add a Ui class to handle interactions with the user

gabriellegtw committed:
Add Storage class

FileReading and FileWriting classes are merged into Storage class

gabriellegtw committed:
Add TaskList and Parser class

Encapsulate commands for adding to the task list to be read and written in the storage class

Parser class encapsulates logic for understanding each line in maxine.txt so that we can remember the tasks created in previous sessions

gabriellegtw committed:
Merge pull request #3 from gabriellegtw/add-gradle-support

Add gradle support

gabriellegtw committed:
Package all files into maxine package

gabriellegtw committed:
Encapsulate parsing into Parser class

Information parsed in input is in Maxine class.

Let's encapsulate this logic into Parser class

gabriellegtw committed:
Add testing for storage and FileParser

gabriellegtw committed:
Create Jar File

gabriellegtw committed:
Add find command

There is no way for users to find tasks based on keywords

A find command is implemented in Parser and Ui classes to enable users to type "find [keywords]" to find a specific or a few specific tasks

gabriellegtw committed:
Merge pull request #4 from gabriellegtw/branch-Level-9

Add find command

gabriellegtw committed:
Align code with coding standards

gabriellegtw committed:
Merge pull request #5 from gabriellegtw/branch-A-CodingStandard

Align code with coding standards

gabriellegtw committed:
Add javadocs to all public methods

gabriellegtw committed:
Merge pull request #6 from gabriellegtw/branch-A-JavaDoc

Add javadocs to all public methods

achl1012 commented on src/main/java/seedu/maxine/FileParser.java:
```diff
@@ -0,0 +1,37 @@
+            return new Deadline(input[2], input[3]);
+        case ("E"):
+            return new Event(input[2], input[3], input[4]);
+        }
```
Good use of the switch statement here! However, the brackets around the 'case' labels are not necessary and can be removed. A default case can be added in as well

achl1012 commented on src/main/java/seedu/maxine/Maxine.java:
```diff
@@ -0,0 +1,62 @@
+     */
+    public static void main(String[] args) {
+        new Maxine().run();
+    }
```
good use of Javadoc comments!

achl1012 commented on src/main/java/seedu/maxine/TaskList.java:
```diff
@@ -0,0 +1,173 @@
+        }
+        Deadline task = new Deadline(desc.toString(), ddl.toString());
+        list.add(task);
+    }
```
for the variables desc and ddl, the full word could be used as the variable name instead to make it more readable

achl1012 commented on src/main/java/seedu/maxine/task/Task.java:
```diff
@@ -0,0 +1,153 @@
+    protected int num;
+    private static int count;
+
+    private static final List<DateTimeFormatter> FORMATS = new ArrayList<>();
```
correct naming for a constant variable here, very good!

achl1012 reviewed (commented):
LGTM. Overall the code mostly follows the code quality guidelines and the Javadoc comments also make the code easier to read and understand, good job!

annabellekk commented on src/main/java/seedu/maxine/Parser.java:
```diff
@@ -0,0 +1,100 @@
+                    list.addDeadline(answer);
+                    storage.refreshStorage(list);
+                } catch (Exception e) {
+                    System.out.println("Please follow this format: deadline [enter maxine.task] /by [enter deadline]");
```
I like how you utilised switch cases for different tasks, the extracting out the different methods with their own function calls, and also providing information in the error messages to help the user pinpoint why an error was thrown and how to fix it

annabellekk commented on src/main/java/seedu/maxine/Storage.java:
```diff
@@ -0,0 +1,81 @@
+
+import seedu.maxine.task.Task;
+
+import java.io.*;
```
Would it be better to list each imported io classes specifically instead of a wildcard import?

annabellekk commented on src/main/java/seedu/maxine/Storage.java:
```diff
@@ -0,0 +1,81 @@
+        File file = new File(filePath);
+
+        try (BufferedWriter writer =
+                     new BufferedWriter(new FileWriter(file, false))) {
```
Should this line be indented with 8 spaces from the try line instead?

annabellekk commented on src/main/java/seedu/maxine/TaskList.java:
```diff
@@ -0,0 +1,173 @@
+    /**
+     * Deletes the task in the collection that corresponds to the key
+     * @param key The number that is tagged to the task that os to be deleted
+     */
```
I like how you included JavaDoc comments for all the public methods, but there should be an empty line between the description and parameters sections. I noticed this issue in several other JavaDoc comments as well.

annabellekk commented on src/main/java/seedu/maxine/Ui.java:
```diff
@@ -0,0 +1,142 @@
+     */
+    public void showLoadingError() {
+        System.out.println("I...can't...load...X_X");
+    }
```
I like this error message, adds some personality to the chatbot 😆

annabellekk reviewed (commented):
Looks good! Not much coding standard violations, just some nits to fix.

gabriellegtw committed:
Copy MainWindow class

Maxine bot lacks a GUI. This makes it difficult for users to use Maxine.

Let's start creating a GUI for users to use Maxine

gabriellegtw committed:
Create GUI for Maxine

Maxine's GUI is not connected to her function

Let's connect it such that we are able to get the correct response from Maxine rather than just an echo of our responses

gabriellegtw committed:
Change Maxine's photo

Maxine's default photo is unsightly

Let's change it for the better

gabriellegtw committed:
Merge pull request #7 from gabriellegtw/branch-Level-10

Change Maxine's photo

gabriellegtw committed:
Add CheckStyle

We should add checkstyle to ensure we are following coding conventions

gabriellegtw committed:
Fix bug for delete

For the delete command, there is a bug that causes Array out of bound exception

Let's fix it so users can successfully delete tasks

gabriellegtw committed:
Shorten add methods in TaskList

The current methods for TaskList is too long

Let's shorten them to improve readability

gabriellegtw committed:
Add assertions

Some methods have certain assumptions. Let's validate them using assertions

gabriellegtw committed:
Encapsulate methods in Parser class

Parse method is more than 50 lines long

Let's split it up further into smaller methods so that the code is more maintainable

gabriellegtw committed:
Add extension (delete all items)

Let's add an extension which allows all tasks to be deleted at once

gabriellegtw committed:
Merge pull request #8 from gabriellegtw/branch-A-Assertions

Add extension (delete all items)

gabriellegtw committed:
Clean up code

gabriellegtw committed:
Add test cases

We don not have a JUnit set up for the Command class - let's implement one!

gabriellegtw committed:
Add more test cases

We don not have a JUnit set up for some methods in the Command class - let's implement more of them!

gabriellegtw committed:
Fix bug for list and bye command

Users are able to use the list and bye command even when the string does not match

gabriellegtw committed:
Tidy up code quality

gabriellegtw committed:
Fix exit command

When we input "bye" into Maxine, we are still able to type in commands following it

gabriellegtw committed:
Fix code quality of Ui class

gabriellegtw committed:
Merge pull request #10 from gabriellegtw/branch-A-CodeQuality

Fix code quality

gabriellegtw committed:
Change some variable names

gabriellegtw committed:
Update README and Ui image

gabriellegtw committed:
Merge pull request #11 from gabriellegtw/Gabrielle-Branch-1

Gabrielle branch 1

gabriellegtw committed:
Update README

gabriellegtw committed:
Update Code Quality

gabriellegtw committed:
Update README.md

gabriellegtw committed:
Update README.md image

gabriellegtw committed:
Update gradle file for JAR

gabriellegtw committed:
Fix Code quality for tests

gabriellegtw committed:
Update file name

Let's edit the file name such that it is not limited to our local machine

gabriellegtw committed:
Update file name

Let's edit the file name such that it is not limited to our local machine

gabriellegtw committed:
Add error handling for delete, mark and unmark

When we type in a number that is greater than the number of items in the task list, an error message prints out in the console but the GUI does not notify the user of this error

Let's handle this error such that the user will be notified of this

gabriellegtw committed:
Update README.md

gabriellegtw committed:
Update README.md

gabriellegtw committed:
Fix typo in README.md

gabriellegtw committed:
Update README.md image

gabriellegtw committed:
Fix code quality in Task class

gabriellegtw committed:
Update README.md

gabriellegtw committed:
Fix formatting in storage class

gabriellegtw committed:
Fix formatting errors in README.md

gabriellegtw committed:
Update README.md
